imports/server/gdriveActivityFetcher.ts

```diff
@@ -0,0 +1,171 @@
+          ts: time,
+          document: document._id,
+          hunt: document.hunt,
+          puzzle: document.puzzle,
+          user: user?._id,
+        });
+      });
+    }
+  }
+}
+
+async function fetchDriveActivity() {
+  const { driveactivity } = GoogleClient;
+  if (!driveactivity) {
+    return;
+  }
+
+  // Don't fetch history that's older than what we'd display
+  const previousTimestamp = Math.max(
+    (await DriveActivityLatests.findOneAsync('default'))?.ts.getTime() ?? 0,
+    Date.now() - ACTIVITY_GRANULARITY * ACTIVITY_SEGMENTS,
+  );
+
+  // Build in some buffer by starting 5 minutes before the latest timestamp
+  // we've previously seen (our unique constraints will dedup any overlap)
+  const filter = `time > ${previousTimestamp - 5 * 60 * 1000}`;
+
+  let pageToken: string | undefined;
+  let latestTimestamp = previousTimestamp;
+  do {
+    const resp = await driveactivity.activity.query({
+      requestBody: {
+        pageToken,
+        filter,
+      },
+    });
+    pageToken = resp.data.nextPageToken ?? undefined;
+
+    if (resp.data.activities) {
+      // Accumulate a promise that resolves to the latest timestamp we've seen
+      latestTimestamp = await resp.data.activities.reduce(async (p, activity) => {
+        const previousLatestTimestamp = await p;
+
+        // See if this is a document edit action
+        if (!activity.actions?.some((action) => action.detail?.edit)) {
+          return previousLatestTimestamp;
+        }
+
+        if (!activity.timestamp || !activity.targets || !activity.actors) {
+          return previousLatestTimestamp;
+        }
+
+        const ts = new Date(activity.timestamp);
+
+        const documentIds = [...activity.targets.reduce<Set<string>>((acc, target) => {
+          if (target.driveItem?.name?.startsWith('items/')) {
+            acc.add(target.driveItem.name.substring('items/'.length));
+          }
+
+          return acc;
+        }, new Set())];
+        const actorIds = [...activity.actors.reduce<Set<string>>((acc, actor) => {
+          if (actor.user?.knownUser?.personName?.startsWith('people/')) {
+            acc.add(actor.user.knownUser.personName.substring('people/'.length));
+          }
+
+          return acc;
+        }, new Set())];
+
+        await recordDriveChanges(ts, documentIds, actorIds);
+
+        return Math.max(previousLatestTimestamp, ts.getTime());
+      }, Promise.resolve(latestTimestamp));
+    }
+
+    pageToken = resp.data.nextPageToken ?? undefined;
+  } while (pageToken);
+
+  await DriveActivityLatests.upsertAsync('default', {
+    $set: {
+      ts: new Date(latestTimestamp),
+    },
+  });
+}
+
+const FEATURE_FLAG_NAME = 'disable.gdrive_document_activity';
+
+async function featureFlagChanged() {
+  return new Promise<void>((r) => {
+    let handle: Meteor.LiveQueryHandle | undefined;
+    const cleanup = () => {
+      handle?.stop();
+      r();
+    };
+    handle = Flags.observeChanges(FEATURE_FLAG_NAME, cleanup);
```

Comment:
Doesn't `Flags.observeChanges` always call the callback with the current value during startup, either because of an `added` or synthesizing an event?  Seems like this will return ~immediately regardless of feature flag state.

Reply:
Yeah, you're right - this structure doesn't work in the way I thought it did. I'll rework it to behave better.

Reply:
I think this is still incorrect?  `featureFlagChanged()`, as written, will never block for more than about a database roundtrip so I expect that if you disabled the feature flag the `while(true)` below would turn into a busy loop and spin up your fans.  I believe that if you want to achieve the "promise blocks until database record changes" semantics, you need to prepare for having the `cleanup` callback not stop the watcher during the initial callback.

I'd suggest having an `let value: boolean | undefined;` outside of the loop, and then in the callback, when value is undefined, you assign it the value of the callback argument (currently unbound), and when value is not undefined, if it changed, you stop the handle and resolve the promise.

Reply:
Yeah, you're right, I had forgotten that I made `observeChanges` emit on initial query to match the behavior of `Cursor.observeChanges`. I did something slightly different, but I think it should work as desired now? (I flipped the flag back and forth a few times to verify that it did not busy-loop)

Reply:
Hmmm, this works now, but I'm a little worried about what happens once we make a future change to make `Flags.active()` -> `await Flags.activeAsync()` since then `initializing` would potentially be set to false again before the async function in the initial `added()` completes.

What you have here won't block the way you want it to once the `Flags.active` call in the `observeChanges` is made non-Fiber-blocking, unless upstream also makes `observeChanges` await each of the callbacks it issues.  I guess we'll notice since we'd have to void the promise, but I worry we'd have new bugs from the ordering-barrier going missing.

I'm not sure how upstream will attempt to handle this, but it seems subtle and like something we can avoid entirely by setting initializing to false in the first triggered callback instead of after the observeChanges.

Possible we'll need to carefully review our usage of observeChanges more broadly as this situation evolves.

Reply:
Hmm. Yeah, I think our strategy here mostly depends on what happens to the behavior of `observeChanges` when they actually switch to an async implementation; there are a few different ways it could go.